subsys/ipc/rpmsg_multi_instance/rpmsg_multi_instance.c

```diff
@@ -0,0 +1,453 @@
+
+	return (IS_ENABLED(CONFIG_RPMSG_MULTI_INSTANCE_MASTER) ?
+			VIRTIO_CONFIG_STATUS_DRIVER_OK :
+			sys_read8(ctx->shm_status_reg_addr));
+}
+
+static uint32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return BIT(VIRTIO_RPMSG_F_NS);
+}
+
+#ifdef CONFIG_RPMSG_MULTI_INSTANCE_MASTER
+
+static void virtio_set_status(struct virtio_device *p_vdev, unsigned char status)
+{
+	struct rpmsg_mi_ctx *ctx  = metal_container_of(p_vdev, struct rpmsg_mi_ctx, vdev);
+
+	sys_write8(status, ctx->shm_status_reg_addr);
+}
+
+
+static void virtio_set_features(struct virtio_device *vdev, uint32_t features)
+{
+	/* No need for implementation */
+}
+#endif
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	int status;
+
+	struct rpmsg_mi_ctx *ctx = metal_container_of(vq->vq_dev, struct rpmsg_mi_ctx, vdev);
+
+	if (ctx) {
+		status = ipm_send(ctx->ipm_tx_handle, 0, 0, NULL, 0);
+		if (status != 0) {
+			LOG_WRN("Failed to notify: %d", status);
+		}
+	}
+}
+
+const static struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.get_features = virtio_get_features,
+#ifdef CONFIG_RPMSG_MULTI_INSTANCE_MASTER
+	.set_status = virtio_set_status,
+	.set_features = virtio_set_features,
+#endif
+	.notify = virtio_notify,
+};
+
+static void ipm_callback_process(struct k_work *item)
+{
+	struct rpmsg_mi_ctx *ctx =
+		CONTAINER_OF(item, struct rpmsg_mi_ctx, ipm_work);
+
+	LOG_DBG("Process callback. Instance name: %s", ctx->name);
+	virtqueue_notification(ctx->vq[VIRTQUEUE_ID]);
+}
+
+static void ipm_callback(const struct device *dev, void *context,
+					 uint32_t id, volatile void *data)
+{
+	ARG_UNUSED(dev);
+	struct rpmsg_mi_ctx *ctx = (struct rpmsg_mi_ctx *)context;
+
+	k_work_submit_to_queue(&ctx->ipm_work_q, &ctx->ipm_work);
+}
+
+int rpmsg_mi_configure_shm(struct rpmsg_mi_ctx *ctx,
+					   const struct rpsmg_mi_ctx_cfg *cfg)
+{
+	uint8_t vring_size  = VRING_SIZE_GET(SHM_SIZE);
+	uint32_t shm_addr   = SHMEM_INST_ADDR_AUTOALLOC_GET(SHM_START_ADDR, SHM_SIZE, instance);
+	uint32_t shm_size   = SHMEM_INST_SIZE_AUTOALLOC_GET(SHM_SIZE);
+
+	ctx->shm_status_reg_addr = shm_addr;
+
+	uint32_t shm_local_start_addr = shm_addr + VDEV_STATUS_SIZE;
+	uint32_t shm_local_size = shm_size - VDEV_STATUS_SIZE;
+
+	ctx->shm_physmap[0] = shm_local_start_addr;
+
+	ctx->shm_device.name = SHM_DEVICE_NAME;
+	ctx->shm_device.bus = NULL;
+	ctx->shm_device.num_regions = 1;
+
+	ctx->shm_device.regions->virt = (void *)shm_local_start_addr;
+	ctx->shm_device.regions->physmap = ctx->shm_physmap;
+
+	ctx->shm_device.regions->size = shm_local_size;
+	ctx->shm_device.regions->page_shift = 0xffffffff;
+	ctx->shm_device.regions->page_mask = 0xffffffff;
+	ctx->shm_device.regions->mem_flags = 0;
+
+	ctx->shm_device.irq_num = 0;
+	ctx->shm_device.irq_info = NULL;
+
+	uint32_t rpmsg_reg_size = 2 * VIRTQUEUE_SIZE_GET(vring_size);
+	uint32_t vring_region_size = VRING_SIZE_COMPUTE(vring_size, VRING_ALIGNMENT);
```

Comment:
declare the variables at the beginning of the function (I was expecting checkpatch to catch this :)

Reply:
Is it a requirement from our coding standard ? I didn't find such requirement here:  https://docs.zephyrproject.org/latest/contribute/index.html#coding-style but I have something like that in my mind.

Reply:
yeah, probably it's just a "nice-to-have" after all.